payments/core.py

```diff
@@ -119,6 +119,22 @@ def get_return_url(self, payment, extra_data=None):
             return url + "?" + qs
         return url
 
+    def autocomplete_with_subscription(self, payment):
+        """
+        Complete the payment with subscription
+        Used by providers, that use server initiated subscription workflow
+
+        Throws RedirectNeeded if there is problem with the payment that needs to be solved by user
+        """
+        raise NotImplementedError()
```

Comment:
+1

It took me several hours to understand what the function is supposed to do. I guess that django-payments rather refers to this as to "making a payment". I would even not hesitate to call it `pay_with_subscription()` (ditto for the `BasePayment` class).

To clarify the docstring even more, I would say that this method is used by _payments_ not _providers_ and that it is used in an _application_-initiated workflow rather than in _server_-initiated workflow as it is not clear what server is meant (I believe I have seen somewhere that django-payments refers to the users of this library as to applications).